Source/Dafny/AST/DafnyAst.cs

```diff
@@ -265,8 +266,9 @@ public class BuiltIns {
     /// the built-in total-arrow type (if "total", in which case "member" is expected to denote the "requires" member).
     /// The given "id" is expected to be already resolved.
     /// </summary>
-    private Expression ArrowSubtypeConstraint(IToken tok, BoundVar id, Function member, List<TypeParameter> tps, bool total) {
+    private Expression ArrowSubtypeConstraint(IToken tok, IToken endTok, BoundVar id, Function member, List<TypeParameter> tps, bool total) {
```

Comment:
The two places in which this `ArrowSubtypeConstraint` is being called actually give `Token.NoToken` as argument, so there is nothing close to `IRegion` to provide, so although this is a good suggestion, it's better to keep it this way until there is a compelling reason to change it.

Reply:
Why not wrap the Tokens in a `class Region : IRegion`?

Reply:
After 1-1 discussion, we agreed that, since Expressions already use `tok` for the start token, it's better to leave this huge refactoring for an independent PR later.

Reply:
We will eventually add start and end positions for every expression in Dafny's grammar, so it's not useful to do this change now in a way that would be half-baked.

Reply:
Seems like there was some confusion, I don't want to approve this PR yet, even though it's merged now.

Could you, for the classes where you added an end position, which already had a start position, replace the existing start position with a region that contains both the start and end position?

Reply:
I honestly don't think of it as an idea that will improve legibility. It's one more level of indirection, duplicates the start token (from `tok`).
Interfaces were created so that an object could display different behaviors. and creating a separate class just to regroup two properties does not seem like a valuable refactoring for now unless we do it on all the expressions.
If we don't do this, we would also need to create also the interface `IHasRegion` to state that an expressions has a `region` field that is itself a `IRegion`. That's one interface too much and I think down the road all expressions should have this region anyway, so that they are initialized with it as you say.

Let's ask for another developer here. @robin-aws , what do you think of this suggested change?

Reply:
@cpitclaudel ?

Reply:
Thanks for the ping, I had a quick look.  I'm not sure *when* the right time is to change the way end tokes are passed, but here's my 2¢:

- Are these tokens used for anything else than position reporting?  If not, they shouldn't even be passed into the various AST constructors, and only ranges instead.  It's OK to implement a range variant that's based on a pair of tokens if we really want to keep the tokens around for some reason.
- Exposing a pair of tokens directly is a risky business, because it's unclear what the semantics of end tokens are.  For example when you convert to a range you use `GetLSPPosition` on both tokens, but that means the range ends on the *beginning* of the last token.  If the last token is part of the range, the range should end at the *end* of that token instead.  If the current code is right, then I expect others will make the same mistake as I did when proofreading this, so I think there's value in a level of indirection that encapsulates pairs of tokens and exposes a range.
- I have no thoughts on *when* that change should be done.  It would be OK to me to put a FIXME in there, but I don't know the context of this change.

HTH!

(Also, I props to you on completing that change; it's impressive me how complicated this conceptually simple change ended up being.)

Reply:
> I have no thoughts on when that change should be done. It would be OK to me to put a FIXME in there, but I don't know the context of this change.

Do we all to agree that the desirable end state is for every expression AST Node to have a Range/Region field instead of a token field? I don't think there is a reason to work with tokens after parsing.

Given that end state, why make a detour by first introducing a second token field to expression AST nodes?

Reply:
I think we all agree on the final state.

It's not a detour, it's actually reusing the same field name and interface as the statements, i.e. being consistent with the system already in place.

Reply:
> it's actually reusing the same field name and interface as the statements, i.e. being consistent with the system already in place.

Okay, let's settle with that then.